submit_cgap/tests/test_submission.py

```diff
@@ -1202,3 +1212,471 @@ def mysterious_error(*args, **kwargs):
+                                                    # bundle_filename argument is not the name of a
+                                                    # metadata bundle file. We did nothing in this mock to
+                                                    # create the file SOME_BUNDLE_FILENAME, so we expect something
+                                                    # like: "The file '/some-folder/foo.xls' does not exist."
+                                                    assert "does not exist" in str(e)
+                                                else:  # pragma: no cover
+                                                    raise AssertionError("Expected ValueError did not happen.")
+
+    # This tests the normal case with validate_only=False and a successful result.
+
+    with shown_output() as shown:
+        with mock.patch("os.path.exists", mfs.exists):
+            with mock.patch("io.open", mfs.open):
+                with io.open(SOME_BUNDLE_FILENAME, 'w') as fp:
+                    print("Data would go here.", file=fp)
+                with mock.patch.object(submission_module, "script_catch_errors", script_dont_catch_errors):
+                    with mock.patch.object(submission_module, "resolve_server", return_value=SOME_SERVER):
+                        with mock.patch.object(submission_module, "yes_or_no", return_value=True):
+                            with mock.patch.object(submission_module, "get_keydict_for_server",
+                                                   return_value=SOME_KEYDICT):
+                                with mock.patch("requests.post", mocked_post):
+                                    with mock.patch("requests.get", make_mocked_get(done_after_n_tries=3)):
+                                        with mock.patch("datetime.datetime", dt):
+                                            with mock.patch("time.sleep", dt.sleep):
+                                                with mock.patch.object(submission_module, "show_section"):
+                                                    with mock.patch.object(submission_module,
+                                                                           "do_any_uploads") as mock_do_any_uploads:
+                                                        try:
+                                                            submit_metadata_bundle(SOME_BUNDLE_FILENAME,
+                                                                                   institution=SOME_INSTITUTION,
+                                                                                   project=SOME_PROJECT,
+                                                                                   server=SOME_SERVER,
+                                                                                   env=None,
+                                                                                   validate_only=False)
```

Comment:
Should this itself be a fixture/context manager? You could also parametrize on the patch values at certain layers, allowing you to test different behavior at each mocked layer.

Reply:
It might be for the cascades that are repeated this isn't a bad idea. Also, the setup of datetime+sleep probably merits a context manager that does that since it's a common idiom.